analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/DoNotHardcodeCredentialsBase.cs

```diff
@@ -106,89 +121,174 @@ protected bool IsEnabled(AnalyzerOptions options)
-            protected abstract string GetVariableName(TSyntaxNode syntaxNode);
-            protected abstract string GetAssignedValue(TSyntaxNode syntaxNode, SemanticModel semanticModel);
+            foreach (var path in context.WebConfigFiles(c))
+            {
+                if (XmlHelper.ParseXDocument(File.ReadAllText(path)) is { } doc)
+                {
+                    CheckWebConfig(c, path, doc.Descendants());
+                }
+            }
+        }
 
-            protected CredentialWordsFinderBase(DoNotHardcodeCredentialsBase<TSyntaxKind> analyzer)
+        private void CheckWebConfig(CompilationAnalysisContext c, string path, IEnumerable<XElement> elements)
+        {
+            foreach (var element in elements)
             {
-                // See https://tools.ietf.org/html/rfc3986 Userinfo can contain groups: unreserved | pct-encoded | sub-delims
-                const string Rfc3986_Unreserved = "-._~";  // Numbers and letters are embeded in regex itself without escaping
-                const string Rfc3986_Pct = "%";
-                const string Rfc3986_SubDelims = "!$&'()*+,;=";
-                const string UriPasswordSpecialCharacters = Rfc3986_Unreserved + Rfc3986_Pct + Rfc3986_SubDelims;
-                var uriUserInfoPart = @"[\w\d" + Regex.Escape(UriPasswordSpecialCharacters) + "]+";
+                if (!element.HasElements && IssueMessage(element.Name.LocalName, element.Value) is { } elementMessage && element.CreateLocation(path) is { } elementLocation)
+                {
+                    c.ReportIssue(Diagnostic.Create(rule, elementLocation, elementMessage));
+                }
+                foreach (var attribute in element.Attributes())
+                {
+                    if (IssueMessage(attribute.Name.LocalName, attribute.Value) is { } attributeMessage && attribute.CreateLocation(path) is { } attributeLocation)
+                    {
+                        c.ReportIssue(Diagnostic.Create(rule, attributeLocation, attributeMessage));
+                    }
+                }
+            }
+        }
 
-                this.analyzer = analyzer;
-                uriUserInfoPattern = new Regex(@"\w+:\/\/(?<Login>" + uriUserInfoPart + "):(?<Password>" + uriUserInfoPart + ")@", RegexOptions.Compiled);
+        private void CheckAppSettings(SonarAnalysisContext context, CompilationAnalysisContext c)
+        {
+            foreach (var path in context.AppSettingsFiles(c))
+            {
+                if (JsonNode.FromString(File.ReadAllText(path)) is { } json)
+                {
+                    CheckAppSettings(c, path, json);
+                }
             }
+        }
 
-            public Action<SyntaxNodeAnalysisContext> AnalysisAction() =>
-                context =>
+        private void CheckAppSettings(CompilationAnalysisContext c, string path, JsonNode json)
+        {
+            var queue = new Queue<JsonNode>();
```

Comment:
Extract as reusable walker?

Reply:
The walker is not as easy at it seemed. I'll leave it like this for now and we can discuss it here and do it in another PR.
The key problem is, that I need to visit `"value"`, but I do not want to visit `"value"` inside object `"key": "value"`. And node doesn't know it's parent.
I don't have a good idea for this stop mechanism without going into recursive descent and enqueuing in the `base.VisitObject`.

How would `Descendants' and `DescendantValues` actually work together?

Reply:
I've drafted attempt in #6201, let's continue the discussion there